Reset Kit Cooldowns

darbyjack opened the pull request:
This is not a finished PR. I'm just opening it so we can keep track of progress.

Mickey42302 committed:
Add Dragon Fireballs to "/fireball".

darbyjack committed:
Started working on kit reset cooldowns.

darbyjack committed:
Started working on kit reset cooldowns.

darbyjack committed:
Added in support for the console to execute this.

darbyjack committed:
Added in support for the console to execute this.

darbyjack committed:
Add in message to tell user kit was reset.

darbyjack committed:
Add in message to tell user kit was reset.

darbyjack commented:
Update: Have given this out for people to test and they've given good feedback.

pop4959 committed:
Implement missing methods in FakeWorld (#2679)

JRoy committed:
Bump supported 1.14.x version to 1.14.4 (#2681)

mdcfe committed:
Fix /r not showing in 1.13+ tab complete

Fixes #2659.

mdcfe commented on Essentials/src/plugin.yml:
```diff
@@ -240,6 +240,10 @@ commands:
+  kitreset:
+    description: Reset the cooldown of a kit for a player.
+    usage: /<command> [player] [kit]
+    aliases: [kr, ekr]
```
Needs to include `/ekitreset`.

darbyjack replied:
Done!

mdcfe committed:
Merge remote-tracking branch 'origin/2.x' into 2.x

darbyjack committed:
Added in /ekitreset to plugin.yml

darbyjack committed:
Added in /ekitreset to plugin.yml

darbyjack committed:
Fixed dupe on new builds. (#2696)

Fixes #2695.

mdcfe committed:
Bump version to 2.17.1

mdcfe committed:
Check plugin equals rather than name equals in autocomplete listener

mdcfe committed:
URGENT: critical bug fix

mdcfe committed:
Implement /delkit command

Closes #2709, closes #2719.

mdcfe committed:
Clean up /essentials

mdcfe committed:
Add aliases to /delkit

mdcfe committed:
Add per-projectile permissions and speed argument to /fireball

triagonal committed:
Implement toggle for command block logging (#2599)

Just a simple config toggle for whether or not command block commands should be logged.

closes #1153, closes #162

---

* Add toggle for command block logs
* update setting name

darbyjack committed:
Check if user has a permission to block putting X item on their head. (#2574)

Closes #2550. 

* Check if user has a permission to block putting X item on their head.
* Forced the material name to be lower case.
* Fixed a typo in the message.
* Updated permission node and added check to the EssentialsPlayerListener.

mdcfe committed:
Implement regex-based nickname blacklist

Adds `nick-blacklist` list to config which takes phrases or regexes. Users with essentials.nick.blacklist.bypass can bypass the blacklist.

mdcfe committed:
Add max-projectile-speed option

Projectils can load chunks pretty intensively if too fast.

mdcfe committed:
Correctly return null for invalid IDs when doing legacy conversion

mdcfe committed:
Fix player head meta on versions below 1.13

mdcfe committed:
Set item meta in setSkullOwner before returning

*Should* address #2765.

mdcfe committed:
Implement missing FakeWorld methods

dashaw92 committed:
Correct unbanip's permission for broadcasts

triagonal committed:
add {USERNAME} placeholder to chat format

triagonal committed:
implement missing OfflinePlayer methods

Ichbinjoe committed:
Save materials in unlimited by name

There was a mistake where materials in unlimited would instead be saved
as a bukkit Material instead of its name. Instead, save it by the name.

rye761 committed:
Make drop-items-if-full work on Essentials kits

rye761 committed:
Implements missing createExplosion methods in FakeWorld (#2818)

darbyjack committed:
Reimplement spawn-if-no-home config option (#2757)

This reimplements the `spawn-if-no-home` config option from the original Essentials plugin. If set to true, running `/home` without having set a home will send the player to spawn (previous behaviour). If set to false, running `/home` will instead show an error message and not teleport the player anywhere.

Closes #1452.

caojohnny committed:
Update Mockito to allow tests to pass in newer versions of Java

mdcfe committed:
Merge pull request #2871 from AgentTroll/patch-mockito

Update Mockito to allow tests to pass in newer versions of Java

JRoy committed:
Update to 1.15 (#2873)

* Compile against 1.15

* Remove achievement related methods for OfflinePlayer

* Add 1.15 to VersionUtil

* Revert "Remove achievement related methods for OfflinePlayer"

This reverts commit d0ab84e0

* Add 1.15 Mobs

* Remove Achievement related methods from OfflinePlayer

This needs more work as this currently breaks > 1.15 support

* Actually this isn't a problem

mdcfe committed:
Add missing FakeWorld methods

Madis0 committed:
Crowdin in readme (#2877)

* Crowdin for readme

* Update README.md

caojohnny committed:
Allow players to set single word AFK message - fixes #2712 (#2876)

Allows players with "essentials.afk.others" to toggle afk themselves with a single word message

caojohnny committed:
Block gamemode changes for jailed players (#2875)

Blocks players from experiencing any gamemode changes while they are jailed.

Fixes #1996.

mdcfe committed:
Build with GitHub Actions (#2800)

* Add GitHub Actions workflow

* Update workflow

* Run BuildTools in workflow

* Update build.yml

* Add distributionManagement to pom.xml

* Add GitHub registry Maven credentials

* Remove Travis CI config

* mkdir -p

* Remove GitHub package registry push

Didn't work anyway

* Remove distributionManagement

* Cache Maven repo

* Fix step order

* Use JitPack for VaultAPI

caojohnny committed:
Correct check for command cooldowns list in userdata

pop4959 committed:
Update to 1.15.1

pop4959 committed:
Add phantoms to protect.prevent.spawn (#2806)

Closes #2733.

mdcfe committed:
Add custom_items.yml file

mdcfe committed:
Add methods to set/remove custom aliases

mdcfe committed:
Fix wrong filename in ManagedFile error message

mdcfe committed:
Remove unused VersionUtil constants

mdcfe committed:
Attempt to parse input as UUID in loop commands (#2606)

Related: #2424

VirtualParticle committed:
Escape pipe character on item lore (#2831)

I just adjusted the regex used to split the item lore into multiple lines in order to allow escaping of the pipe character using \|.

Fixes #2830.

JRoy committed:
Add a cause enum to UserBalanceUpdateEvent (#2824)

Basically, I just needed this for one of my plugins; otherwise, this could be useful for people using this event who want to see where the user's balance is being updated from.

---

* Add UserBalanceUpdateEvent.Cause
* Add special cause enum
* Add API Cause

VirtualParticle committed:
Enable control of entity transformations in EssentialsProtect (#2836)

This allows users to prevent any of the following transformations:

- Creeper charging
- Villager infected by zombie villagers
- Villager being cured
- Villagers turning into witches
- Pigs turning into zombie pigmen
- Zombies turning into Drowned (and husks turning into Zombies)
- Mooshrooms switching colors

Configurations are in a new subsection, `protect.prevent.transformation`. All options are disabled by default.

---

* Allow for control of transformations

* Undid removal of villager checking

* Undid removal of villager checking

* Documentation for transformation configs

Co-Authored-By: md678685 <1917406+md678685@users.noreply.github.com>

* Removed duplicate configuration lines

* Updated info about husks drowning

* Changed config names

Co-authored-by: md678685 <1917406+md678685@users.noreply.github.com>

mdcfe committed:
Fix custom_items.yml generation

mdcfe committed:
Don't assume primary group is always present

Fixes a potential issue with PEX 2.0 (and 1.x) where users may not have a primary group.
Closes #2901.

mdcfe committed:
Minor refactor of permissions handlers

* Move `essentials.build` perm check from AntiBuild into default SuperPerms handler
* Don't emulate wildcards for LuckPerms or PEX
* Add `build: true` meta support for LuckPerms (+ other plugins as needed)

mdcfe committed:
Slightly nicer AntiBuild messages

CmdrKittens committed:
Don't attempt to modify a Map value view (#2910)

md dun goofed

montlikadani committed:
Improve scheduler methods to use Java 1.8 lambdas (#2700)

* Update EssentialsPlayerListener.java

* Update Backup.java

pop4959 committed:
Improvements for delkit (#2908)

3 improvements for delkit within this PR:

1) Cleaned up unused imports
2) Show actual kit name when deleting instead of memory address (this is the reason originally for making this)
3) Implement tab completion

I've tested to make sure the changes are properly functioning. See below:

Before:
![2019-12-29_01 12 09](https://user-images.githubusercontent.com/17698576/71555204-4bd5c200-29de-11ea-9fe3-eb88db08ef48.png)

After:
![2019-12-29_01 48 50](https://user-images.githubusercontent.com/17698576/71555207-56905700-29de-11ea-9277-271bb0656d64.png)
![2019-12-29_01 48 25](https://user-images.githubusercontent.com/17698576/71555208-58f2b100-29de-11ea-8257-742fa7bb3150.png)

montlikadani committed:
Update PlayerList.java

montlikadani committed:
Update Commandlightning.java

pop4959 committed:
Update FakeWorld again (#2925)

Spigot added some new methods.

pop4959 committed:
Update GeoIP (#2926)

This implements the fix suggested by @Bobcat00 in #2919

Users of GeoIP will now be required to register a MaxMind account and create a license to download the database required by the plugin. This license is entered into the new license-key field in the configuration.

---

* Update GeoIP

* Simplify branch

* Update old config detection

* Update geoIpLicenseMissing to point to the EssX wiki

mdcfe committed:
Fix Java 9 breakage causing issues on Java 8

pop4959 committed:
Fix bed spawn message on 1.15+ (#2943)

Closes #2893

**Tests**
![2020-01-13_14 08 12](https://user-images.githubusercontent.com/17698576/72297425-be66b480-3610-11ea-8421-1ac6881259b8.png)
![2020-01-13_14 10 02](https://user-images.githubusercontent.com/17698576/72297431-c32b6880-3610-11ea-89fc-bdda2d0cec89.png)

pop4959 committed:
Properly register entity transform event depending on version (#2942)

Fixes the problem mentioned in #2931 which was introduced in #2836. This is caused as a result of EntityTransformEvent being a relatively new event introduced in Spigot [856adc3e6af](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/commits/856adc3e6afc3e9c0495ac27e38b96b144e54b1e), which was after the update to 1.13.2 in Spigot [f311182396b](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/commits/f311182396bd74b106efebc926ea4d0b92cfba80).

Closes #2931.

pop4959 committed:
Bump supported version to 1.15.2

mdcfe committed:
Split plugin compatibility report into own class

pop4959 committed:
Update FakeServer (#2997)

imDaniX committed:
Properly escape pipe character on item lore (#2994)

Using negative lookbehind instead of negated characters set.

Fixes #2962, fixes #2984

triagonal committed:
Implement missing FakeWorld methods (#3002)

mdcfe committed:
Revert "Split plugin compatibility report into own class"

Revert 6fa5b80c0a16bf02d474ddf383e8d4709caf72e1, which broke `/ess version`, until this can be fixed.

mdcfe committed:
Update funding to include Ko-fi link

triagonal committed:
Add beezooka command (#3003)

This PR introduces the `/beezooka` command as suggested in #2975. The command acts similarly to `/kittycannon` with the exception of the Bee being tamed, instead flying around for a small amount of time before exploding.

Sample usage:
![beezooka](https://user-images.githubusercontent.com/10545540/74099459-11804880-4b78-11ea-9c3d-499b06d912fb.gif)

As Bees do not exist below version 1.15, an error will display when attempting to run the command on such a version:
![image](https://user-images.githubusercontent.com/10545540/74099477-4d1b1280-4b78-11ea-91c2-fa675994a85f.png)

Closes #2975.

JasonHorkles committed:
Explain how to add commands to kits (#2980)

* Explain how to add commands to kits

* Update Essentials/src/kits.yml

Co-Authored-By: md678685 <1917406+md678685@users.noreply.github.com>

Co-authored-by: md678685 <1917406+md678685@users.noreply.github.com>

JRoy committed:
Fix CustomItemResolver causing a StackOverflowError (#3022)

CustomItemResolver was fetching the custom item from the item database instead of the resolved custom item which caused a StackOverflowError.

Fixes #3018.

mdcfe committed:
Add 1.14/1.15 item aliases and fallbacks

Short term replacement for #2897 until ItemDbGenerator is complete:

* Closes #2934 by adding 1.15 items
* Closes #2790 by adding minecraft:sign fallbacks and a "sign" alias
* Closes #2732 by adding aliases and fallbacks for cactus green/green dye, dandelion yellow/yellow dye and rose red/red dye.

JasonHorkles committed:
Update config.yml group-formats (#3025)

Since most (if not all) modern permission plugins only use lowercase group names, the examples should also be lowercase. A note has also been added explaining that group names are case-sensitive.

Commit:
Clarify easter egg config description (#3028)

mdcfe committed:
Reduce permission check calls in PlayerCommandSendEvent (#3026)

This reduces permissions checks by only checking each command once, not checking each command for each alias.

The impact of this is a near-tenfold reduction in permission checks during this listener, which should slightly improve performance when using LuckPerms (and provide significant improvements on older inefficient permissions plugins).

mdcfe committed:
Fix /fireball on 1.8.8 (#3027)

Dragon fireballs were added in 1.9 and so this previously caused an
error. This PR fixes that (and adds tridents as throwables). Fixes #3016.

mdcfe committed:
Bump version to 2.17.2

\o/

pop4959 committed:
Add missing OfflinePlayer methods (#3033)

montlikadani committed:
Properly fix error with fireball command (#3040)

Closes #3016.

* Update Commandfireball.java

* Update Commandfireball.java

Co-authored-by: md678685 <1917406+md678685@users.noreply.github.com>

mdcfe committed:
Fix bee spawn egg aliases

pop4959 committed:
Near exclude permission

pop4959 committed:
Fix tab completion for Exp command

pop4959 committed:
Fix zh-CN GeoIP locale issue

pop4959 committed:
Update FakeWorld + FakeServer (#3061)

JRoy committed:
Implement better AfkStatusChangeEvent

JRoy committed:
Fix User::payUser(User, BigDecimal, Cause) not properly utilizing Cause (#3064)

werogg committed:
Add speed info to whois (#2952)

darbyjack committed:
Added in option to remove effects on heal (#2754)

montlikadani committed:
Add tab-completer for /spawnmob command (#2661)

darbyjack committed:
Fixed double local logging (#2755)

pop4959 committed:
Implement max mute time feature (#2807)

montlikadani committed:
Add {PLAYER} placeholder with uppercase in kits (#2929)

triagonal committed:
Revert "Add {PLAYER} placeholder with uppercase in kits (#2929)" (#3072)

This reverts commit 90c4a2f53cc6f05f14c65c45310b191ad3acd53f.

pop4959 committed:
Update FakeWorld (#3095)

Markyroson committed:
Make MockBukkit compatible #3105

Pokechu22 committed:
Mention 1.15.2 in the README (#3107)

pop4959 committed:
Use isAuthorized for permission check

pop4959 committed:
Fix bed message on some older versions of MC (#3048)

Fixes #3045. Some older versions of MC use "BED_BLOCK" to represent beds.

Commit:
permission-based jailed player interactions (#2503)

Commit:
implement command teleporting to a player's last known logout location (#2505)

* implement /offlinetp command, teleports to a player's last known logout location

* send a message to /tp user when they try teleporting to an offline player

* getHidden is now false, you are no longer able to teleport to offline players using offline tp

* change /offlinetp to /tpoffline to match essentialsx conventions

montlikadani committed:
Add teleport warmup event to API (#2590)

* Update Teleport.java

* Create UserTeleportEvent.java

* Update Teleport.java

* Add location to event

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/commands/Commandkitreset.java:
```diff
@@ -0,0 +1,67 @@
+        List<String> options = new ArrayList<>();
+        if (args.length == 1 && user.isAuthorized("essentials.kit.reset")) {
+            return getPlayers(server, user);
+        } else if (args.length == 2)  {
```
Missing second essentials.kit.reset permission check?

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/commands/Commandkitreset.java:
```diff
@@ -0,0 +1,67 @@
+            }
+            return options;
+        }
+        return Collections.emptyList();
```
return options;, you already constructed it, why not use it!

darbyjack replied:
I think I rebased this correctly. Added the changes.

montlikadani committed:
Implement message for #2640  (#2721)

* Update Essentials.java

* Update messages.properties

* Update messages.properties

Xeyame committed:
Add different translation keys for the "me" message (#2805)

* Add different translation keys for the "me" message

* Add proper German "me" translation

Ichbinjoe committed:
Make drop-items-if-full work on Essentials kits (#2820)

Ichbinjoe committed:
Merge pull request #2914 from montlikadani/patch-5

Fix NPE when the user is null in the hidden players

Ichbinjoe committed:
Merge pull request #3049 from pop4959/near-exempt

Near exclude permission

Ichbinjoe committed:
Merge pull request #3051 from pop4959/xp-tab-complete-fix

Fix tab completion for Exp command

Ichbinjoe committed:
Merge pull request #3052 from pop4959/fix-zh-CN-geoip-issue

Fix zh-CN GeoIP locale issue

darbyjack committed:
Fixed my stupid PR

Ichbinjoe committed:
Merge pull request #3063 from schlatt-co/feature/3034

Implement better AfkStatusChangeEvent

Ichbinjoe committed:
Merge pull request #3111 from darbyjack/200-IQ-Protect-Improvements

200 IQ Essentials Protect Improvements

darbyjack committed:
Version to 1.14

Ichbinjoe committed:
Merge pull request #3113 from darbyjack/200-IQ-Protect-Improvements

Fixed my PR pt2

Ichbinjoe committed:
Merge pull request #3106 from Markyroson/2.x

Make MockBukkit compatible #3105

darbyjack committed:
Added second perm check and returned options.

darbyjack committed:
Merge remote-tracking branch 'origin/feature/resetkitcommand' into feature/resetkitcommand

darbyjack committed:
Stupid git

Ichbinjoe commented:
lmao this pr is fucked
